amaranth/hdl/_time.py

```diff
@@ -0,0 +1,249 @@
+import numbers
+import re
+
+
+__all__ = ["Period"]
+
+
+class Period:
+    def __init__(self, **kwargs):
+        if not kwargs:
+            self._femtoseconds = 0
+            return
+
+        if len(kwargs) > 1:
+            raise TypeError("Period accepts at most one argument")
+
+        (unit, value), = kwargs.items()
+
+        m = re.match(r"^([munpf]?)s|([kMG]?)Hz$", unit)
```

Comment:
using a regex here is unnecessarily opaque, and also duplicates information already contained in the dicts below.

I'd do something like that:

```
TIME_UNITS = {
    "s": 1_000_000_000_000_000,
    "ms": 1_000_000_000_000,
    # ...
}

FREQUENCY_UNITS = {
    "Hz": 1_000_000_000_000_000,
    "kHz": 1_000_000_000_000,
    # ...
}

if unit in TIME_UNITS:
    self._femtoseconds = round(value * TIME_UNITS[unit])
elif unit in FREQUENCY_UNITS:
    # ...
    self._femtoseconds = round(FREQUENCY_UNITS[unit] / value)
else:
    raise TypeError(...)
```

Reply:
I like Wanda's approach better.